Cache ly_ctx_info to improve get performance

alangefe opened the pull request:
Fixes #304

This change results in a 3x to 4x speedup for unfiltered get requests (and for requests for the ietf-yang-library contents).

We now cache the results of the `ly_ctx_info(np2srv.ly_ctx)` call and save it in the `np2srv` global, to avoid repeating the call when it would just return the same data.

We update the cached value when necessary, namely, when a module or feature is changed.

Because the contents of the cache should change relatively infrequently, it seems worth it to pay the cost early instead of for every request.

michalvasko commented:
Hi Andrew,
you are right and we have nothing against this optimization. However, I will request some changes before I will merge it.

Firstly, there is an [identificator](https://tools.ietf.org/html/rfc7895#section-2.1.1) that changes whenever the context changes so you can easily learn whether your cached info is up-to-date by comparing this value (*libyang* `ly_ctx_get_module_set_id()`), which should simplify your code.

Secondly, you certainly get much better response time for `<get>` but this data tree can be quite large so you must not forget about higher memory requirements. These are exactly the kind of optimizations we are recently working recently on, but because of the trade-off, we make them optional. So, I would like you to add a `cmake` variable `ENABLE_CACHE` which would internally translate to defining `NP2_ENABLED_CACHE` (look into *libyang*, that is what I have in mind). It can be `ON` by default.

Thank you if you implement it this way and if not, then I will do it once I have some spare time.

Regards,
Michal

alangefe commented:
Thanks for the feedback; I've addressed your concerns in the latest commit.

I decided to use the more descriptive flag name `ENABLE_LY_CTX_INFO_CACHE` instead of the generic `ENABLE_CACHE`, to avoid giving the impression that other data was being cached. (At the moment, we're really only caching the results from `ly_ctx_info()`.)

michalvasko commented:
Hi,
I asked you to name it `ENABLE_CACHE` because I expect there will be more of such optimizations in future but we can rename it then, no problem. Other than that it is OK, merging.

Regards,
Michal
